rodlayout/proxy.py

```diff
@@ -1,18 +1,218 @@
+        self, sep: TupleVector = (0, 0), maintain: bool = False, **handle: '_AlignHandle'
+    ) -> 'FigureCollection':
+        """
+        Aligns this figure on a reference figure:
+
+        >>> instance.pr_boundary.align(center_left=ref_figure.b_box.center_right)
+
+        Returns a figure collection containing all shapes involved in the alignment.
+        """
+        assert len(handle) == 1, "Exactly 1 align handle expected."
+        key, ref = handle.popitem()
+        assert ref._handle, "Need a handle name for the reference object."
+        try:
+            align_handle = python_skill_handle[key]
+        except KeyError:
+            raise ValueError(f"There is no handle named {key!r}.") from None
+        return self._do_align(align_handle=align_handle, ref=ref, sep=sep, maintain=maintain)
+
+    def _do_align(
+        self, align_handle: str, ref: '_AlignHandle', sep: TupleVector, maintain: bool
+    ) -> 'FigureCollection':
+        if maintain:
+            raise ValueError("Cannot maintain alignment when figure has no rod representation.")
+        cell_view = self._shape.cell_view
+        with ghost_shape(cell_view, self._shape, self._pr) as align_rod:
+            with ghost_shape(cell_view, ref._shape, ref._pr) as ref_rod:
+                current_workspace.rod.align(
+                    align_obj=align_rod,
+                    align_handle=align_handle,
+                    ref_obj=ref_rod,
+                    ref_handle=ref._handle,
+                    maintain=False,
+                    x_sep=sep[0],
+                    y_sep=sep[1],
+                )
+        return FigureCollection([self._shape, ref._shape])
+
+
+@dataclass(frozen=True)
+class _RodAlignHandle(_AlignHandle):
+    """
+    The rod align handle tries to align without usage of the ghost_shape trick.
+    This is possible if aligned on bounding boxes and no figure collection is involved.
+    """
+
+    def _do_align(
+        self, align_handle: str, ref: '_AlignHandle', sep: TupleVector, maintain: bool
+    ) -> 'FigureCollection':
+        if maintain and not isinstance(ref._shape, RodShape):
+            raise ValueError(
+                "Cannot maintain alignment when ref objects has no rod representation."
+            )
+        if maintain and (self._pr or ref._pr):
+            raise ValueError("Cannot maintain alignment for pr boundary.")
+
+        if (
+            not isinstance(ref._shape, RodShape)
+            or not isinstance(self._shape, RodShape)
+            or self._pr
+            or ref._pr
+        ):
+            return super()._do_align(align_handle, ref, sep, maintain)
+
+        current_workspace.rod.align(
+            align_obj=self._shape.rod,
+            align_handle=align_handle,
+            ref_obj=ref._shape.rod,
+            ref_handle=ref._handle,
+            maintain=maintain,
+            x_sep=sep[0],
+            y_sep=sep[1],
+        )
+        return FigureCollection([self._shape, ref._shape])
+
```

Comment:
This two classes should be in a separated file

Reply:
Since the align handles are deeply interleaved with the Figure class, I see no clear logical separation to put in an extra file.